### PR TITLE
Use sbt-github-actions to github actions CI

### PR DESCRIPTION
**Pull request by mdedetrich**
This PR adds https://github.com/sbt/sbt-github-actions which is slowly becoming a standard amongst Scala sbt github projects for automatically generating a github actions workflow from an SBT build. The advantage of sbt-github-actions is that it reduces the maintenance burden of a project, i.e. if you add/update a scala version and then run `sbt githubWorkflowGenerate` it will handle everything for you and make sure that its in sync.

For now we are using https://github.com/sbt/sbt-github-actions just for CI however its possible to also use it in conjunction with sbt-ci-release in order to make releases (see https://github.com/sbt/sbt-github-actions#integration-with-sbt-ci-release). This would allow you to make a proper release just by pushing a git tag to the git repo and it can also handle automatic snapshot publishing whenever a PR is merged into `master`.

Another added bonus of sbt-github-actions is it sets up a proper matrix which speeds up the build (since it parallelizes it) and gives better info if a specific Scala/JVM version fails. This is in conjunction to use using caching.

Note that the reason this PR cannot be merged right now is because there is a pre-existing github actions check called `test (17)` which no longer applies. This means the check needs to be removed, merge the  PR and aftewards the new checks can be added for whatever matrix you want.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -9,6 +9,9 @@ val isScala3 = Def.setting(
   CrossVersion.partialVersion(scalaVersion.value).exists(_._1 == 3)
 )
 
+ThisBuild / scalaVersion       := Scala3
```
Comment (mdedetrich) [inline build.sbt:12]: This had to be done for technical sbt reasons, basically sbt-github-actions will only pick up these values if they are in the `ThisBuild` scope. In general the concept of `commonSettings` is outdated for newer sbt builds (thats why `ThisBuild` was invented) but changing the entire `build.sbt` would be for another PR.

File: build.sbt [hunk 2]
```diff
@@ -251,3 +252,25 @@ lazy val parboiledCore = crossProject(JSPlatform, JVMPlatform, NativePlatform)
     Compile / sourceGenerators += generateActionOps.taskValue
   )
   .nativeSettings(nativeSettings)
+
+ThisBuild / githubWorkflowTargetTags ++= Seq("v*")
+ThisBuild / githubWorkflowPublishTargetBranches := Seq()
```
Comment (mdedetrich) [inline build.sbt:257]: This disables publishing for now, if you want to use sbt-github-actions for publishing just let me know and I will set it up in another PR.

File: build.sbt [hunk 3]
```diff
@@ -251,3 +252,25 @@ lazy val parboiledCore = crossProject(JSPlatform, JVMPlatform, NativePlatform)
...
+
+ThisBuild / githubWorkflowJavaVersions := List(JavaSpec.temurin("8"), JavaSpec.temurin("11"), JavaSpec.temurin("17"))
```
Comment (mdedetrich) [inline build.sbt:259]: One important thing to note here is that the first value in this list is whats used for publishing (if we decide to do it). I am not sure what JVM parboiled2 is published under, but if its not JDK 8 I need to update this.

**Comment by sirthias** [comment]
Thank you, Matthew, for your PRs with the awesome explanations and background info!
Top-notch stuff!

> This PR adds https://github.com/sbt/sbt-github-actions which is slowly becoming a standard amongst Scala sbt github projects ...

Ok, very interesting! Thank you!

> For now we are using https://github.com/sbt/sbt-github-actions just for CI however its possible to also use it in conjunction with sbt-ci-release in order to make releases...

I see. At this time I don't think automatic releases add any real value for this project here.
They are simply too infrequent. I'd assume that properly automating and maintaining the process takes longer than all release work of the next 5 years... :)

File: build.sbt (hunk 1 above)
Comment (sirthias) [reply build.sbt:12]: > In general the concept of commonSettings is outdated for newer sbt builds...

Ok, thank you! It seems I'm a bit out of touch with the state-of-the-art.
It's time I look a bit through more modern sources to see how things are done nowadays.
Therefore your comment are very much appreciated, thank you again!

File: build.sbt (hunk 3 above)
Comment (sirthias) [reply build.sbt]: JDK 8 should be fine at this time. Thank you!

**Comment by mdedetrich** [comment]
> I see. At this time I don't think automatic releases add any real value for this project here.
They are simply too infrequent. I'd assume that properly automating and maintaining the process takes longer than all release work of the next 5 years... :)

It would take 5-10 minutes, its literally only a few lines of code (basically https://github.com/sbt/sbt-github-actions#integration-with-sbt-ci-release). If you don't want to do it thats also fine

**Comment by yanns** [comment]
I confirm, it's quite easy.
The secrets have to be added to github to be used like: https://github.com/sangria-graphql/sangria/blob/main/build.sbt#L27-L30
Once it's done, the release process is just pushing a tag: https://github.com/sangria-graphql/sangria/blob/main/HOW-TO-RELEASE.md

The benefit for me is to control completely the environment where the artifacts are being build.
I once pushed a new release inadvertently built with a new JVM that broke some other projects.

**Comment by mdedetrich** [comment]
To demonstrate how easy it is I already setup a PR for at it https://github.com/sirthias/parboiled2/pull/431, the only thing thats needed is to add the sonatype/pgp github secrets.

> I once pushed a new release inadvertently built with a new JVM that broke some other projects.

Indeed this is the main advantage, the release environment is completely controlled and aside from pushing the wrong version in the git tag its very hard to screw up a release.

**Comment by sirthias** [comment]
Ok, gentlemen, apparently you know how to push my buttons... :)
I just merged #431 and hopefully set everything up correctly on the secrets side.
Fingers crossed... :)
